Kaa 773 log delivery callbacks

forGGe opened the pull request:
Includes:
- implementation
- tests
- small improvements to the test framework
- couple of nitpicks

forGGe committed:
KAA-773 Add interfaces, stubs and documentation

forGGe committed:
KAA-773 Extend test framework with useful macros

Now it is possible to setup per-test fixture based
on group in which it included.

forGGe committed:
KAA-773 Add basic log delivery test

Commit:
KAA-773 renamed timeout info in bucket info; set log delivery listeners

Commit:
KAA-773 save record count in remember_request

Commit:
KKA-773 return expected bucket id

Commit:
KAA-773 Call listeners for success, timeout and fail

forGGe committed:
KAA-773 Clarify log record interfaces

forGGe committed:
KAA-773 Implement log count summary for listeners

forGGe committed:
KAA-773 Extend test facilities for CUnit case

In order to support per-test setup()/teardown() procedures

forGGe committed:
KAA-773 Small cleanup

forGGe committed:
KAA-773 Extend test cases with timeouts, context and ID checks

Test structure improved as well

rasendubi commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -53,25 +53,30 @@ typedef enum {
+typedef struct
+{
+    uint16_t     log_bucket_id;     /**< ID of bucket present in storage. */
+    kaa_time_t   timeout;           /**< bucket timeout */
```
move timeout to start to avoid padding

forGGe replied:
Done

rasendubi commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -263,15 +282,17 @@ static void update_storage(kaa_log_collector_t *self)
+        KAA_LOG_ERROR(self->logger,
+                      KAA_ERR_BADDATA, "Failed to add log record: "
+                                       "serialized record size is null. "
+                                       "Maybe log record schema is empty");
```
It's better to either move `KAA_ERR_BADDATA` to first line, or move string to a separate line.

forGGe replied:
Done

rasendubi commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
+typedef struct
+{
+    uint16_t bucket_id;     /**< ID of bucket present in storage. */
+    size_t   log_count;     /**< Logs left to upload across all buckets. */
```
move log_count one line up. That doesn't change anything now, but has better chance to avoid padding in the future.

forGGe replied:
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
+ * @brief Log bucket information structure.
+ * One or more log records aggregated into the bucket.
+ */
+typedef struct
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
+/**
+ * @brief Event handler type.
+ *
+ * Bucket information can be used to retrieve a amount of logs that still
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
+ * @param[in,out]  ctx    User-definied context. @sa kaa_logging_add_record
+ * @param[in]      bucket Log bucket for which event was triggered.
+ */
+typedef void (*kaa_log_event_fn)(void *ctx, const kaa_log_bucket_info_t *bucket);
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
+    kaa_log_event_fn on_failed;  /**< Handler called upon failed delivery. */
+    kaa_log_event_fn on_timeout; /**< Handler called upon timeouted delivery. */
+    void *ctx;                   /**< User-defined context. */
+} kaa_log_listeners_t;
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.h:
```diff
@@ -42,6 +42,54 @@ extern "C" {
     typedef struct kaa_log_collector        kaa_log_collector_t;
 #endif
 
+/**
```
Done

forGGe committed:
KAA-773 Core review comments addressed

forGGe committed:
KAA-773 Move size constraints from strategy to logging modules

Patch contains fixes for warnings from tests and API usage in various
places as well.

forGGe committed:
KAA-773 Add log strategy and log storage setters

Tests are included

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -34,45 +34,44 @@
+    bool                           is_sync_ignored;
+    uint32_t                       log_last_id;         /**< Last log record ID */
+    uint32_t                       max_bucket_log_count;
+    uint32_t                       max_storage_size;
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -297,14 +342,29 @@ kaa_error_t kaa_logging_add_record(kaa_log_collector_t *self, kaa_user_log_recor
         update_storage(self);
 
+    if (!self->log_bucket_id) {
+        self->log_bucket_id = self->status->log_bucket_id;
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/platform-impl/Econais/EC19D/econais_ec19d_kaa_client.c:
```diff
@@ -91,7 +91,8 @@ void print_mem_stat(kaa_client_t *kaa_client);
 #define KAA_DEMO_MAX_UPLOAD_THRESHOLD     15   /* Size of collected serialized logs needed to initiate log upload */
-#define KAA_DEMO_MAX_LOG_BUCKET_SIZE      16   /* Max size of a log batch has been sent by SDK during one upload. */
+#define KAA_DEMO_MAX_LOG_BUCKET_SIZE      512   /* Max size of a log batch has been sent by SDK during one upload. */
+#define KAA_DEMO_MAX_LOGS_IN_BUCKET       16   /* Max count of logs in one bucket */
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -34,45 +34,44 @@
-
-
+#define KAA_MAX_LOGS_IN_BUCKET             64
+#define KAA_MAX_BUCKET_SIZE                1024
```
Done.

forGGe commented on client/client-multi/client-c/src/kaa/platform-impl/esp8266/esp8266_kaa_client.c:
```diff
@@ -60,6 +60,10 @@ static kaa_service_t OPERATIONS_SERVICES[] = { KAA_SERVICE_PROFILE
 
+/* Logging constraints */
+#define MAX_LOG_COUNT           SIZE_MAX
+#define MAX_LOG_BUCKET_SIZE     (8 * 1024)
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/platform-impl/posix/posix_kaa_client.c:
```diff
@@ -67,6 +67,9 @@ static kaa_service_t OPERATIONS_SERVICES[] = { KAA_SERVICE_PROFILE
 
+/* Logging constraints */
+#define MAX_LOG_COUNT           SIZE_MAX
+#define MAX_LOG_BUCKET_SIZE     (8 * 1024)
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/platform-impl/stm32/leafMapleMini/esp8266/esp8266_kaa_client.c:
```diff
@@ -70,7 +70,9 @@ static kaa_service_t OPERATIONS_SERVICES[] = { KAA_SERVICE_PROFILE
-
+/* Logging constraints */
+#define MAX_LOG_COUNT           SIZE_MAX
+#define MAX_LOG_BUCKET_SIZE     (2 * 1024)
```
Done.

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -463,9 +527,29 @@ kaa_error_t kaa_logging_handle_server_sync(kaa_log_collector_t *self
+                          sum_log_records, &total_logs);
+
+        kaa_log_bucket_info_t log_bucket_info = {
+            .log_count = total_logs,
```
Done

forGGe committed:
KAA-773 Code review comments addressed

forGGe committed:
KAA-773 Fix log count retrieval

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -104,21 +105,32 @@ static kaa_error_t remember_request(kaa_log_collector_t *self, uint16_t bucket_i
+    struct find_bucket_context *find_context = context;
 
+    if (((timeout_info_t *)data)->log_bucket_id == find_context->bucket_id) {
+        find_context->bucket_count++;
```
Done

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -34,45 +34,45 @@
 
+struct find_bucket_context {
+    uint16_t bucket_id;
+    size_t   bucket_count;
```
Done.

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -104,21 +105,32 @@ static kaa_error_t remember_request(kaa_log_collector_t *self, uint16_t bucket_i
+
+/* Returns amount of items processed if positive.
+ * Or error if negative */
+static ssize_t remove_request(kaa_log_collector_t *self, uint16_t bucket_id)
```
Done.

forGGe commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -455,17 +516,38 @@ kaa_error_t kaa_logging_handle_server_sync(kaa_log_collector_t *self
+
+        ssize_t count = remove_request(self, bucket_id);
+
+        if (count < 0) {
```
Done.

ohamov commented on client/client-multi/client-c/src/kaa/kaa_logging.c:
```diff
@@ -104,24 +100,32 @@ static kaa_error_t remember_request(kaa_log_collector_t *self, uint16_t bucket_i
+    timeout_info_t *timeout_to_find = context;
+    timeout_info_t *timeout_info = data;
 
+    if (timeout_info->log_bucket_id == timeout_to_find->log_bucket_id) {
```
To avoid context state changes, please use **kaa_list_find_next()** to get the necessary info and **kaa_list_remove_at()** to remove the element

forGGe committed:
KAA-773 Improve timeouts list traversal
